pkg/virt-api/rest/subresource.go

```diff
@@ -194,7 +195,16 @@ func (app *SubresourceAPIApp) getVirtHandlerConnForVMI(vmi *v1.VirtualMachineIns
 	if !vmi.IsRunning() {
 		return nil, goerror.New(fmt.Sprintf("Unable to connect to VirtualMachineInstance because phase is %s instead of %s", vmi.Status.Phase, v1.Running))
 	}
-	return kubecli.NewVirtHandlerClient(app.virtCli).Port(app.consoleServerPort).ForNode(vmi.Status.NodeName), nil
+
+	nodeName := vmi.Status.NodeName
+	if cachedConnection, exists := app.virtHandlerCons[nodeName]; exists {
+		return cachedConnection, nil
+	}
+
+	virtHandlerCon := kubecli.NewVirtHandlerClient(app.virtCli).Port(app.consoleServerPort).ForNode(nodeName)
```

Comment:
You can pre-create on http client as part of the `SubresourceAPIApp` initialization and always pass this http client as an additional paramter to `app.NewVirtHandlerClient(app.virtCli, app.httpCli)`. This client can already use the right auto-updating tls config.

You then don't need the  tls parameter on the Get and Put calls anymore.

Also you don't need any caching per node. Just ose this one http client for everything.

Reply:
Done!!
Thanks a lot for the suggestion